test/util/dynamic/dynamic.go

```diff
@@ -0,0 +1,89 @@
+package dynamic
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// Client returns a client from a given unstructured object.
+// It can be used when creating an object from a yaml file.
+type Client interface {
+	GetClient(obj *unstructured.Unstructured) (ResourceClient, error)
+}
+
+type client struct {
+	dynamic.Interface
+	mapping meta.RESTMapper
+}
+
+func NewDynamicClient(kubeConfig *rest.Config) (Client, error) {
+	cli, err := dynamic.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	groupResources, err := restmapper.GetAPIGroupResources(discoveryClient)
+	if err != nil {
+		return nil, err
+	}
+	mapper := restmapper.NewDiscoveryRESTMapper(groupResources)
+
+	return &client{cli, mapper}, nil
+}
+
+func (d *client) getGVR(obj *unstructured.Unstructured) (*meta.RESTMapping, error) {
+	var gvk = obj.GroupVersionKind()
+	return d.mapping.RESTMapping(gvk.GroupKind(), gvk.Version)
+}
+
+// GetClient returns a non-namespaced or namespaced ResourceClient depending on a given object.
+func (d *client) GetClient(obj *unstructured.Unstructured) (ResourceClient, error) {
+	gvr, err := d.getGVR(obj)
+	if err != nil {
+		return nil, err
+	}
+	if obj.GetNamespace() != "" {
```

Comment:
nit: if we want to be perfectly correct, we need to use discovery to determine if the GVR is namespaced or not - a user passing in a namespaced object but forgetting to set a namespace will cause a subtle error here